### PR TITLE
Adding optional hip test

### PR DESCRIPTION
**Pull request by afzpatel**
Installation command: 
**spack install hip**

Command to run test:
**spack test run hip**

The console output is shown in hip_test_console_out.txt
[hip_test_console_out.txt](https://github.com/spack/spack/files/10759360/hip_test_console_out.txt)

The full test log is shown in hip-5.4.0-kmk3ovf-test-out.txt
[hip-5.4.0-kmk3ovf-test-out.txt](https://github.com/spack/spack/files/10759381/hip-5.4.0-kmk3ovf-test-out.txt)

There are some examples that are failing which are displayed at the bottom of the test log(hip-5.4.0-kmk3ovf-test-out.txt).  The console output will show a failure for the entire test even if one example fails.

### CODE REVIEW

**Review by srekolam — APPROVED** [review]

**Comment by renjithravindrankannath** [comment]
@afzpatel , test run failed in centos8, ubuntu18 and 20 with 5.4.3. Please check the log below. 
[hip-5.4.3-centos8-svsx637-test-out.txt](https://github.com/spack/spack/files/11457053/hip-5.4.3-centos8-svsx637-test-out.txt)
[hip-5.4.3-ubuntu18.04-2lesgc3-test-out.txt](https://github.com/spack/spack/files/11457055/hip-5.4.3-ubuntu18.04-2lesgc3-test-out.txt)
[hip-5.4.3-ubuntu20.04-ucr3fhj-test-out.txt](https://github.com/spack/spack/files/11457056/hip-5.4.3-ubuntu20.04-ucr3fhj-test-out.txt)

**Review by renjithravindrankannath — APPROVED** [review]

**Review by renjithravindrankannath — APPROVED** [review]

**Comment by afzpatel** [comment]
@haampie @tldahlgren this pr is ready to be reviewed

File: var/spack/repos/builtin/packages/hip/package.py [hunk 1]
```diff
@@ -120,7 +120,7 @@ class Hip(CMakePackage):
 
     depends_on("cuda", when="+cuda")
 
-    depends_on("cmake@3.16.8:", type="build", when="@4.5.0:")
+    depends_on("cmake@3.16.8:", type=("build", "run"), when="@4.5.0:")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/hip/package.py]: Is `run` being added here so it is available for the tests?  If so, you should be able to instead set `test_requires_compiler` to `True` then pick up the right version of `cmake` through `self.spec["cmake"].command` or `which(self.spec["cmake"].prefix.bin.cmake)`. 

(See note about `cmake` at https://spack.readthedocs.io/en/latest/packaging_guide.html#enabling-test-compilation.)

File: var/spack/repos/builtin/packages/hip/package.py [hunk 2]
```diff
@@ -606,3 +607,68 @@ def cmake_args(self):
...
+        install test subdirectory for use during `spack test run`."""
+        if self.spec.satisfies("@:5.1.0"):
+            return
+        self.cache_extra_test_sources([self.test_src_dir])
+
+    def test_samples(self):
+        # configure, build and run all hip samples
+        if self.spec.satisfies("@:5.1.0"):
+            print("Skipping: stand-alone tests")
+            return
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/hip/package.py]: Looks like I need to update the documentation.  In the meantime, you could change this to raise the new `SkipTest` exception here, along the lines:

```suggestion
        if self.spec.satisfies("@:5.1.0"):
            raise SkipTest("Test is only available for specs after version 5.1.0")
```

This is the new way of flagging a test case as being skipped for CDash (so CI) reporting.

File: var/spack/repos/builtin/packages/hip/package.py [hunk 3]
```diff
@@ -606,3 +607,68 @@ def cmake_args(self):
...
+                                and not file.endswith(".o")
+                            ):
+                                print("Executing test binary: " + file)
+                                exe = which(file)
+                                if file == "hipDispatchEnqueueRateMT":
+                                    options = ["16", "0"]
+                                else:
+                                    options = []
+                                exe(*options)
+                        make("clean")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/hip/package.py]: Curious.  Is this needed since the tests are running in the test stage directory?

**Review by tldahlgren — COMMENTED** [review]
I mainly have some questions and an optional suggestion.  I am currently trying to determine why the tests are failing for me.

File: var/spack/repos/builtin/packages/hip/package.py [hunk 4]
```diff
@@ -606,3 +607,68 @@ def cmake_args(self):
...
+                    ):
+                        if "CMakeLists.txt" in files:
+                            print("Configuring  test " + test_name)
+                            os.mkdir("build")
+                            os.chdir("build")
+                            cmake(*cc_options)
+
+                        print("Building test " + test_name)
+                        make(parallel=False)
+                        # itterate through the files in dir to find the newly built binary
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/hip/package.py]:
```suggestion
                        # iterate through the files in dir to find the newly built binary
```

**Comment by afzpatel** [comment]
> I mainly have some questions and an optional suggestion. I am currently trying to determine why the tests are failing for me.

Implemented suggestions. Also, one test named 22_cmake_hip_lang is currently failing. We are tracking this failure.

**Comment by tldahlgren** [comment]
> > I mainly have some questions and an optional suggestion. I am currently trying to determine why the tests are failing for me.
> 
> Implemented suggestions. Also, one test named 22_cmake_hip_lang is currently failing. We are tracking this failure.

Thanks.  I must not be building it correctly (`spack install hip`) because, on the system I'm using with the default options, only 7 tests pass.  Since it's likely "user error" I won't let that stop this PR.

FYI, I get:
```
$ spack -v test run hip
==> Spack test 3c2xc3zvxlr32elogguaverxix6qij7k
==> Testing package hip-5.4.3-jmksfdm
==> [2023-07-27-16:13:20.775417] Installing $SPACK/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hip-5.4.3-jmksfdm42r4edcpbbwvmy2laf2rtfwsy/.spack/test to $SPACK_TEST_ROOT/3c2xc3zvxlr32elogguaverxix6qij7k/hip-5.4.3-jmksfdm/cache/hip
==> [2023-07-27-16:13:22.153822] test: test_samples: unspecified purpose
==> [2023-07-27-16:13:22.163292] test: test_sample_bit_extract: configure, build and run test: bit_extract
Configuring  test bit_extract
...
PASSED: Hip::test_samples
==> [2023-07-27-16:18:18.263964] Completed testing
==> [2023-07-27-16:18:18.264365] 
========================== SUMMARY: hip-5.4.3-jmksfdm ==========================
Hip::test_sample_bit_extract .. FAILED
Hip::test_sample_module_api .. PASSED
Hip::test_sample_module_api_global .. FAILED
Hip::test_sample_square .. FAILED
Hip::test_sample_hipBusBandwidth .. PASSED
Hip::test_sample_hipCommander .. FAILED
Hip::test_sample_hipDispatchLatency .. FAILED
Hip::test_sample_hipInfo .. FAILED
Hip::test_sample_0_MatrixTranspose .. FAILED
Hip::test_sample_10_inline_asm .. FAILED
Hip::test_sample_11_texture_driver .. FAILED
Hip::test_sample_12_cmake_hip_add_executable .. FAILED
Hip::test_sample_13_occupancy .. PASSED
Hip::test_sample_14_gpu_arch .. PASSED
Hip::test_sample_device_functions .. FAILED
Hip::test_sample_host_functions .. FAILED
Hip::test_sample_16_assembly_to_executable .. FAILED
Hip::test_sample_17_llvm_ir_to_executable .. FAILED
Hip::test_sample_18_cmake_hip_device .. FAILED
Hip::test_sample_19_cmake_lang .. FAILED
Hip::test_sample_1_hipEvent .. FAILED
Hip::test_sample_20_hip_vulkan .. FAILED
Hip::test_sample_21_cmake_hip_cxx_clang .. FAILED
Hip::test_sample_22_cmake_hip_lang .. FAILED
Hip::test_sample_23_cmake_hiprtc .. PASSED
Hip::test_sample_3_shared_memory .. FAILED
Hip::test_sample_4_shfl .. FAILED
Hip::test_sample_5_2dshfl .. FAILED
Hip::test_sample_6_dynamic_shared .. FAILED
Hip::test_sample_7_streams .. PASSED
Hip::test_sample_8_peer2peer .. FAILED
Hip::test_sample_9_unroll .. FAILED
Hip::test_samples .. PASSED
======================= 26 failed, 7 passed of 33 parts ========================
==> [2023-07-27-16:18:18.264420] 
...
```

In a number of cases I see a Make warning:
```
CMake Warning (dev) at CMakeLists.txt:21 (project):
  cmake_minimum_required() should be called prior to this top-level project()
  call.  Please see the cmake-commands(7) manual for usage documentation of
  both commands.
This warning is for project developers.  Use -Wno-dev to suppress it.
```

**Review by tldahlgren — DISMISSED** [review]
Thanks for the updates.

**Review by renjithravindrankannath — APPROVED** [review]

**Review by tldahlgren — APPROVED** [review]

**Pull request merged**